File: compilers/lab1/lexer.go
package main

import "unicode"

type Lexer struct {
	input    string
	position int
	length   int
}

func NewLexer(input string) *Lexer {
	return &Lexer{
		input:    input,
		length:   len(input),
		position: 0,
	}
}

// Peek - смотрит на следующий символ без изменения позиции
func (l *Lexer) peek(offset int) rune {
	peekPos := l.position + offset
	if peekPos < l.length {
		return rune(l.input[peekPos])
	}
	return '\000'
}

// Next - возвращает текущий символ и продвигает позицию
func (l *Lexer) next() rune {
	if l.position >= l.length {
		return '\000'
	}
	current := rune(l.input[l.position])
	l.position++
	return current
}

func (l *Lexer) Tokenize() []Token {
	var tokens []Token
	for l.position < l.length {
		current := rune(l.input[l.position])
		if unicode.IsSpace(current) {
			l.position++
		} else if unicode.IsDigit(current) {
			tokens = append(tokens, l.readNumber())
		} else if unicode.IsLetter(current) || current == '_' {
			tokens = append(tokens, l.readIdentifierOrKeyword())
		} else if current == '"' {
			tokens = append(tokens, l.readString())
		} else {
			tokens = append(tokens, l.readSymbol())
		}
	}
	tokens = append(tokens, NewToken(EOF, "", l.position))
	return tokens
}

func (l *Lexer) readNumber() Token {
	start := l.position
	for l.position < l.length && unicode.IsDigit(rune(l.input[l.position])) {
		l.position++
	}
	value := l.input[start:l.position]
	return NewToken(NUMBER, value, start)
}

func (l *Lexer) readIdentifierOrKeyword() Token {
	start := l.position
	for l.position < l.length && (unicode.IsLetter(rune(l.input[l.position])) || unicode.IsDigit(rune(l.input[l.position])) || l.input[l.position] == '_') {
		l.position++
	}
	value := l.input[start:l.position]
	switch value {
	case "var":
		return NewToken(VAR, value, start)
	case "print":
		return NewToken(PRINT, value, start)
	case "if":
		return NewToken(IF, value, start)
	case "else":
		return NewToken(ELSE, value, start)
	case "while":
		return NewToken(WHILE, value, start)
	default:
		return NewToken(ID, value, start)
	}
}

func (l *Lexer) readString() Token {
	start := l.position
	l.position++ // Skip opening quote
	for l.position < l.length && l.input[l.position] != '"' {
		if l.input[l.position] == '\\' && l.position+1 < l.length {
			l.position++ // Skip escape character
		}
		l.position++
	}
	if l.position < l.length {
		l.position++ // Skip closing quote
	}
	value := l.input[start:l.position]
	return NewToken(STRING, value, start)
}

func (l *Lexer) readSymbol() Token {
	start := l.position
	current := rune(l.input[l.position])
	l.position++

	// Проверка двухсимвольных операторов
	switch current {
	case '=':
		if l.peek(0) == '=' {
			l.position++
			return NewToken(EQEQ, "==", start)
		}
		return NewToken(EQ, "=", start)

	case '!':
		if l.peek(0) == '=' {
			l.position++
			return NewToken(NEQ, "!=", start)
		}
		return NewToken(EXCL, "!", start)

	case '<':
		if l.peek(0) == '=' {
			l.position++
			return NewToken(LTEQ, "<=", start)
		}
		return NewToken(LT, "<", start)

	case '>':
		if l.peek(0) == '=' {
			l.position++
			return NewToken(GTEQ, ">=", start)
		}
		return NewToken(GT, ">", start)

	case '&':
		if l.peek(0) == '&' {
			l.position++
			return NewToken(AND, "&&", start)
		}
		return NewToken(UNKNOWN, "&", start)

	case '|':
		if l.peek(0) == '|' {
			l.position++
			return NewToken(OR, "||", start)
		}
		return NewToken(UNKNOWN, "|", start)

	case '+':
		return NewToken(PLUS, "+", start)
	case '-':
		return NewToken(MINUS, "-", start)
	case '*':
		return NewToken(STAR, "*", start)
	case '/':
		return NewToken(SLASH, "/", start)
	case '(':
		return NewToken(LPAREN, "(", start)
	case ')':
		return NewToken(RPAREN, ")", start)
	case '{':
		return NewToken(LBRACE, "{", start)
	case '}':
		return NewToken(RBRACE, "}", start)
	case ';':
		return NewToken(SEMICOLON, ";", start)
	default:
		return NewToken(UNKNOWN, string(current), start)
	}
}

File: compilers/lab1/Lexer/Lexer/Lexer.cs
using System.Collections.Generic;

namespace Lexer
{
    public class Lexer
    {
        private readonly string input;
        private int position;
        private readonly int length;

        public Lexer(string input)
        {
            this.input = input;
            this.length = input.Length;
            this.position = 0;
        }

        // Peek - смотрит на следующий символ без изменения позиции
        private char Peek(int offset = 1)
        {
            int peekPos = position + offset;
            return peekPos < length ? input[peekPos] : '\0';
        }

        // Next - возвращает текущий символ и продвигает позицию
        private char Next()
        {
            if (position >= length)
                return '\0';
            return input[position++];
        }

        // AddToken - добавляет токен в список
        private void AddToken(List<Token> tokens, TokenType type, string value, int startPos)
        {
            tokens.Add(new Token(type, value, startPos));
        }
        public List<Token> Tokenize()
        {
            var tokens = new List<Token>();
            while (position < length)
            {
                char current = input[position];
                if (char.IsWhiteSpace(current))
                {
                    position++;
                }
                else if (char.IsDigit(current))
                {
                    tokens.Add(ReadNumber());
                }
                else if (char.IsLetter(current) || current == '_')
                {
                    tokens.Add(ReadIdentifierOrKeyword());
                }
                else if (current == '"')
                {
                    tokens.Add(ReadString());
                }
                else
                {
                    tokens.Add(ReadSymbol());
                }
            }
            tokens.Add(new Token(TokenType.EOF, "", position));
            return tokens;
        }
        private Token ReadNumber()
        {
            int start = position;
            while (position < length && char.IsDigit(input[position]))
            {
                position++;
            }
            string value = input.Substring(start, position - start);
            return new Token(TokenType.NUMBER, value, start);
        }
        private Token ReadIdentifierOrKeyword()
        {
            int start = position;
            while (position < length && (char.IsLetterOrDigit(input[position]) || input[position] == '_'))
            {
                position++;
            }
            string value = input.Substring(start, position - start);
            return value switch
            {
                "var" => new Token(TokenType.VAR, value, start),
                "print" => new Token(TokenType.PRINT, value, start),
                "if" => new Token(TokenType.IF, value, start),
                "else" => new Token(TokenType.ELSE, value, start),
                "while" => new Token(TokenType.WHILE, value, start),
                _ => new Token(TokenType.ID, value, start)
            };
        }
        private Token ReadString()
        {
            int start = position;
            position++; // Skip opening quote
            while (position < length && input[position] != '"')
            {
                if (input[position] == '\\' && position + 1 < length)
                {
                    position++; // Skip escape character
                }
                position++;
            }
            if (position < length)
            {
                position++; // Skip closing quote
            }
            string value = input.Substring(start, position - start);
            return new Token(TokenType.STRING, value, start);
        }

        private Token ReadSymbol()
        {
            int start = position;
            char current = input[position];
            position++;

            // Проверка двухсимвольных операторов
            switch (current)
            {
                case '=':
                    if (Peek(0) == '=')
                    {
                        position++;
                        return new Token(TokenType.EQEQ, "==", start);
                    }
                    return new Token(TokenType.EQ, "=", start);

                case '!':
                    if (Peek(0) == '=')
                    {
                        position++;
                        return new Token(TokenType.NEQ, "!=", start);
                    }
                    return new Token(TokenType.EXCL, "!", start);

                case '<':
                    if (Peek(0) == '=')
                    {
                        position++;
                        return new Token(TokenType.LTEQ, "<=", start);
                    }
                    return new Token(TokenType.LT, "<", start);

                case '>':
                    if (Peek(0) == '=')
                    {
                        position++;
                        return new Token(TokenType.GTEQ, ">=", start);
                    }
                    return new Token(TokenType.GT, ">", start);

                case '&':
                    if (Peek(0) == '&')
                    {
                        position++;
                        return new Token(TokenType.AND, "&&", start);
                    }
                    return new Token(TokenType.UNKNOWN, "&", start);

                case '|':
                    if (Peek(0) == '|')
                    {
                        position++;
                        return new Token(TokenType.OR, "||", start);
                    }
                    return new Token(TokenType.UNKNOWN, "|", start);

                case '+':
                    return new Token(TokenType.PLUS, "+", start);
                case '-':
                    return new Token(TokenType.MINUS, "-", start);
                case '*':
                    return new Token(TokenType.STAR, "*", start);
                case '/':
                    return new Token(TokenType.SLASH, "/", start);
                case '(':
                    return new Token(TokenType.LPAREN, "(", start);
                case ')':
                    return new Token(TokenType.RPAREN, ")", start);
                case '{':
                    return new Token(TokenType.LBRACE, "{", start);
                case '}':
                    return new Token(TokenType.RBRACE, "}", start);
                case ';':
                    return new Token(TokenType.SEMICOLON, ";", start);
                default:
                    return new Token(TokenType.UNKNOWN, current.ToString(), start);
            }
        }
    }
}